app/src/main/java/io/appium/settings/RecorderThread.java

```diff
@@ -0,0 +1,367 @@
+
+import android.annotation.SuppressLint;
+import android.hardware.display.DisplayManager;
+import android.hardware.display.VirtualDisplay;
+import android.media.AudioAttributes;
+import android.media.AudioFormat;
+import android.media.AudioPlaybackCaptureConfiguration;
+import android.media.AudioRecord;
+import android.media.MediaCodec;
+import android.media.MediaCodecInfo;
+import android.media.MediaFormat;
+import android.media.MediaMuxer;
+import android.media.projection.MediaProjection;
+import android.os.Handler;
+import android.util.DisplayMetrics;
+import android.util.Log;
+import android.view.Surface;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+public class RecorderThread implements Runnable {
+
+    private static final String TAG = "RecorderThread";
+
+    private MediaProjection mediaProjection;
+    private VirtualDisplay virtualDisplay;
+    private Surface surface;
+    private MediaCodec audioEncoder;
+    private MediaCodec videoEncoder;
+    private AudioRecord audioRecord;
+    private MediaMuxer muxer;
+    private boolean muxerStarted;
+    private boolean startTimestampInitialized;
+    private long startTimestampUs;
+    private Handler handler;
+    private Thread audioRecordThread;
+
+    private int videoTrackIndex = -1;
+    private int audioTrackIndex = -1;
+
+
+    private String outputFilePath;
+    private String videoMime;
+    private int videoWidth;
+    private int videoHeight;
+    private int videoBitrate;
+    private int frameRate;
+    private int timeLapse;
+    private int sampleRate;
+    private Thread recordingThread;
+
+    private volatile boolean stopped = false;
+    private volatile boolean audioStopped;
+    private volatile boolean asyncError = false;
+
+    @SuppressLint("NewApi")
+    private VirtualDisplay.Callback displayCallback = new VirtualDisplay.Callback() {
+        @Override
+        public void onPaused() {
+            super.onPaused();
+            Log.v(TAG, "Display paused");
+        }
+
+        @Override
+        public void onStopped() {
+            super.onStopped();
+            if (!stopped) {
+                asyncError = true;
+            }
+        }
+    };
+
+    public RecorderThread(MediaProjection mediaProjection, String outputFilePath,
+                          int videoWidth, int videoHeight, int videoBitrate) {
+        this.mediaProjection = mediaProjection;
+        handler = new Handler();
+        this.outputFilePath = outputFilePath;
+        this.videoWidth = videoWidth;
+        this.videoHeight = videoHeight;
+        this.videoBitrate = videoBitrate;
+        this.sampleRate = 44100;
+        this.timeLapse = 1;
+        this.frameRate = 30;
+        videoMime = MediaFormat.MIMETYPE_VIDEO_AVC;
+    }
+
+    public void startRecording() {
+        stopped = false;
+        recordingThread = new Thread(this);
+        recordingThread.start();
+    }
+
+    public void stopRecording() {
+        stopped = true;
```

Comment:
should recordingThread be stopped/nullified here?

Reply:
we are just calling this function on RecorderService stopScreenRecord() method and with the latest update, we now nullifying recorderThread object